### ISSUE TITLE
[proposal] Way to export/display the convolution algorithm found with cudnn.benchmark = True

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#52439]
In the future it may be good to allow to export these choices to allow onnxruntime-gpu use exact same conv params and not have surprising slow-downs (https://github.com/microsoft/onnxruntime/issues/6744)

Full implementation of this feature may require some standard for exchanging these parameters between frameworks, maybe the ONNX format could include these in some metadata section

**Comment by spandantiwari** [comment]
@vadimkantorov -  The PyTorch exporter and also the ONNX standard has to be largely backend agnostic. I am wondering if this request is specific to CUDA/CuDNN. If there's a clear way to standardize this across most backends then it may be possible to bring something like this in the ONNX standard. Feel free to send any proposal to ONNX.

**Comment by vadimkantorov** [comment]
@spandantiwari The main focus of this proposal is being able to access somehow algorithm choices made by PyTorch. Of course, such execution hints are specific to backend. In more general context, allowing the user to control certain operations may be interesting as well (e.g. there may be different backends for SVD and providing the choice to the user makes sense). At first, I think some explicit arguments/parameters `method` or `algorithm` for the ops would be enough.

I think first some hacky ways would be already good to prove its usefuleness (e.g. special method in PyTorch and special method in onnxruntime), standardization can come later.

**Comment by ptrblck** [comment]
> I think first some hacky ways would be already good to prove its usefuleness (e.g. special method in PyTorch and special method in onnxruntime), standardization can come later.

I don't think this is the right approach, as it might make the proposed "cudnn algo selection" unusable.

While I think we can add a debugging mechanism to show which cudnn algos, workspace sizes etc. were selected, note that frameworks might be using a different approach of calling cudnn workloads. E.g. one framework might disable non-deterministic kernels completely, while these might be picked in PyTorch via `cudnn.benchmark=True`. Besides that, the algo selection is also depending on the used cudnn version, hardware etc.
I don't know how onnxruntime calls cudnn and if `cudnnFind` is used, which seems to be the root cause of the performance difference.

**Comment by vadimkantorov** [comment]
I guess then the first step should be just being able to access/trace the made algo choices then. cudnn logging can be too verbose (as it also includes benchmarking calls), more user-friedliness could be added e.g. wrt were tensor cores used or not

**Comment by vadimkantorov** [comment]
@ptrblck Indeed it seems to call `cudnnFindConvolutionForwardAlgorithmEx` if that's what you meant by `cudnnFind`

**Comment by vadimkantorov** [comment]
@ptrblck Indeed, ONNXRT is buggy wrt `cudnnFindConvolutionForwardAlgorithmEx` usage: https://github.com/microsoft/onnxruntime/issues/7212

But coming up with a standard syntax for algo choice/hints may still be beneficial for PyTorch, as sometimes the user may want to control the underlying choices with a standard way across the framework

**Comment by vadimkantorov** [comment]
A simlar issue on op-scoped execution hints control: https://github.com/pytorch/pytorch/issues/83392

**Comment by vadimkantorov** [comment]
Related:
- https://github.com/pytorch/pytorch/issues/116369

This might be analogous to TRT engine functionality which saves to file chosen benchmarked alto choices and then lets the user use them
